analyses/chromosomal-instability/util/circos-plots.R

```diff
@@ -0,0 +1,338 @@
+      palette_col <- RColorBrewer::brewer.pal(ifelse(n_levels < 3, 3, n_levels),
+        name = palette
+      )
+
+      # Make a recode color key list
+      level_key <- palette_col[1:n_levels]
+
+      # Make the palette colors the names in this list
+      names(level_key) <- levels(bed_df$color_key)
+
+      # Set up the color ready data.frame
+      bed_df <- bed_df %>%
+        # Make column that specifies the color for each value
+        dplyr::mutate(
+          color_key = dplyr::recode(color_key, !!!level_key),
+          color_key = as.character(color_key)
+        )
+
+      # Only keep the colors that were actually recoded
+      palette_col <- unique(bed_df$color_key)
+    }
+  } else {
+    # If no color key is specified, just make the color black.
+    palette_col <- "black"
+  }
+
+  # Check if the chromosomes have the `chr` format. If not, add it.
+  if (!grepl("chr", bed_df$chr[1])) {
+    bed_df <- bed_df %>%
+      dplyr::mutate(chr = paste0("chr", chr))
+  }
+
+  # If we aren't adding on a track to a current plot...
+  if (!add_track) {
+    # Initialize the plot
+    circlize::circos.clear()
+    circlize::circos.par(start.degree = 90)
+    circlize::circos.initializeWithIdeogram(species = "hg38")
+  }
+
+  if (type == "point") {
+    # Circlize wants things in their own column for each color
+    bed_df <- bed_df %>%
+      # Make a color column and then spread to each so circlize can use it
+      tidyr::spread(color_key, y_val) %>%
+      dplyr::select(-rowid)
+
+    # Add scatterplot track
+    circlize::circos.genomicTrackPlotRegion(bed_df,
+      track.height = track_height,
+      # Establish the ylim
+      ylim = c(y_min, y_max),
+      panel.fun = function(region, value, ...) {
+        circlize::circos.genomicPoints(region, value,
+          col = palette_col
+        )
+      }
+    )
+  } else if (type == "line") {
+    # For line and point, circlize wants things in their own column for each color
+    bed_df <- bed_df %>%
+      # Make a color column and then spread to each so circlize can use it
+      tidyr::spread(color_key, y_val) %>%
+      dplyr::select(-rowid)
+
+    # Add line track
+    circlize::circos.genomicTrackPlotRegion(bed_df,
+      track.height = track_height,
+      # Establish the ylim
+      ylim = c(y_min, y_max),
+      panel.fun = function(region, value, ...) {
+        circlize::circos.genomicLines(region, value,
+          col = palette_col
+        )
+      }
+    )
+  } else if (type == "rect") {
+    # Get rid of rowid
+    bed_df <- bed_df %>%
+      dplyr::select(-rowid)
+    # Add a rectangle track
+    circlize::circos.genomicTrackPlotRegion(bed_df,
+      track.height = track_height,
+      # Establish the ylim
+      ylim = c(y_min, y_max),
+      panel.fun = function(region, value, ...) {
+        i <- circlize::getI(...)
+        circlize::circos.genomicRect(region, value[[1]],
+          ytop = value[[1]] + 0.4,
+          ybottom = value[[1]] - 0.4,
```

Comment:
Should we make the rect size an option for this function? So +-0.4 would be derived from an argument, rather than hard coded here.

Reply:
Yes, this was one of those things, along with many other parameters that should probably be its own option, but I wasn't sure which of the option ideas I had to prioritize.